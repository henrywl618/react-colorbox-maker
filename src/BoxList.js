import React, {useState} from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from 'uuid';


const BoxList = ()=> {

    const startingBoxes = [{width:"100px",height:"100px",color:"green", id:uuid()},
                           {width:"75px",height:"75px",color:"blue", id:uuid()}]
    let [boxes, setBoxes] = useState(startingBoxes);

    const addNewBox = (formData)=>{
        setBoxes(currentBoxes => (
            [...currentBoxes,{...formData, id:uuid()}]
        ));
    }; 

    const removeBox = (id) => {
        setBoxes(currentBoxes => {
            let copy = currentBoxes.filter(box => box.id !== id);
            return copy
        });
    };

    return(
        <>
        <NewBoxForm addNewBox={addNewBox}/>
        {boxes.map(box => <Box width={box.width} height={box.height} color={box.color} id={box.id} removeBox={()=>removeBox(box.id)} />)}
        </>
    );
};

export default BoxList;