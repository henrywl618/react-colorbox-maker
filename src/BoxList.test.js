import React from "react";
import BoxList from "./BoxList";
import {fireEvent, render} from "@testing-library/react";

it("renders without crashing", ()=>{
    render(<BoxList />)
});

it("matches the snapshot", ()=>{
    const {asFragment}=render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", ()=>{
    const {getByLabelText, queryByText, queryAllByText} = render(<BoxList />);

    //there should only be 2 boxes prepopulated
    let boxButtons = queryAllByText("X");
    expect(boxButtons.length).toBe(2);

    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const colorInput = getByLabelText("Color:");
    const submitBtn = queryByText("Submit");

    fireEvent.change(widthInput, {target:{value:"100px"}});
    fireEvent.change(heightInput, {target:{value:"100px"}});
    fireEvent.change(colorInput, {target:{value:"yellow"}});
    fireEvent.click(submitBtn);

    boxButtons = queryAllByText("X");
    expect(boxButtons.length).toBe(3);

});

it("can delete a box", ()=>{
    const {queryAllByText} = render(<BoxList />);

    //there should only be 2 boxes prepopulated
    let boxButtons = queryAllByText("X");
    expect(boxButtons.length).toBe(2);

    fireEvent.click(boxButtons[0]);

    boxButtons = queryAllByText("X");
    expect(boxButtons.length).toBe(1);

    fireEvent.click(boxButtons[0]);

    boxButtons = queryAllByText("X");
    expect(boxButtons.length).toBe(0);

});