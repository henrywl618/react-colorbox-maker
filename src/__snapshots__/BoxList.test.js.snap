// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="text"
      value=""
    />
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="text"
      value=""
    />
    <br />
    <button>
      Submit
    </button>
  </form>
  <div
    class="Box"
    style="width: 100px; height: 100px; background-color: green;"
  />
  <button
    class="Box-button"
  >
    X
  </button>
  <div
    class="Box"
    style="width: 75px; height: 75px; background-color: blue;"
  />
  <button
    class="Box-button"
  >
    X
  </button>
</DocumentFragment>
`;
